share/sysbuild/cmake/modules/sysbuild_extensions.cmake

```diff
@@ -351,6 +368,93 @@ function(ExternalZephyrProject_Cmake)
+#
+# This function defines groups of external Zephyr projects. Images added using
+# `ExternalZephyrProject_Add()` may inherit common properties of the group(s)
+# they are assigned to.
+#
+# This function can either add a new group "<name>", or extend the definition of
+# an existing group with that name. This lets multiple sources influence what a
+# given group should represent in the multiimage build system. The final set of
+# images/properties for every group is evaluated in `sysbuild_groups_resolve()`.
+#
+# <name>                 Name of the group.
+# INCLUDE <images>:      Add a set of images to the group.
+# EXCLUDE <images>:      Remove a set of images from the group.
+# INCLUDE_FROM <groups>: INCLUDE all images belonging to each group in the list.
+# EXCLUDE_FROM <groups>: EXCLUDE all images belonging to each group in the list.
+# IMMUTABLE:             Disallow extending the definition of this group.
+#                        Raises an error if the group already exists.
+#
+# Example use:
+#   ExternalZephyrProject_Group(g_1
+#                               INCLUDE i_1 i_2 i_3 i_4
+#   )
+#   ExternalZephyrProject_Group(g_2
+#                               INCLUDE_FROM g_1
+#                               EXCLUDE i_1 i_2
+#   )
+#   ExternalZephyrProject_Group(g_3
+#                               INCLUDE_FROM g_1
+#                               EXCLUDE_FROM g_2
+#   )
+#   ExternalZephyrProject_Group(g_4
+#                               INCLUDE_FROM g_3
+#                               EXCLUDE_FROM g_1
+#   )
+#
+# In this example, the contents of each group will be as follows:
+#   g_1: i_1, i_2, i_3, i_4
+#   g_2: i_3, i_4
+#   g_3: i_1, i_2
+#   g_4: <empty>
+#
+function(ExternalZephyrProject_Group name)
+  set(options IMMUTABLE)
+  set(multi_args INCLUDE INCLUDE_FROM EXCLUDE EXCLUDE_FROM)
+  cmake_parse_arguments(ZGROUP "${options}" "" "${multi_args}" ${ARGN})
+  if(ZGROUP_UNPARSED_ARGUMENTS)
+    message(FATAL_ERROR
+      "ExternalZephyrProject_Group(${ARGV0} <val> ...) given unknown arguments:"
+      " ${ZGROUP_UNPARSED_ARGUMENTS}"
+    )
+  endif()
+
+  set(target_name ${name}_group)
+  if(NOT TARGET ${target_name})
+    add_custom_target(${target_name})
+    set_property(GLOBAL APPEND PROPERTY sysbuild_groups ${name})
+  elseif(ZGROUP_IMMUTABLE)
+    message(FATAL_ERROR
+      "ExternalZephyrProject_Group(${ARGV0} <val> ...) failed to define "
+      "group '${name}' as IMMUTABLE, because it already exists."
+    )
+  else()
+    get_property(is_immutable TARGET ${target_name} PROPERTY IMMUTABLE)
+    if(is_immutable)
+      message(FATAL_ERROR
+        "ExternalZephyrProject_Group(${ARGV0} <val> ...) group '${name}' is "
+        "IMMUTABLE and cannot be extended."
+      )
+    endif()
+  endif()
+
+  set_property(TARGET ${target_name} PROPERTY IMMUTABLE "${ZGROUP_IMMUTABLE}")
+  foreach(arg ${multi_args})
+    if(DEFINED ZGROUP_${arg})
+      set_property(TARGET ${target_name} APPEND PROPERTY ${arg} "${ZGROUP_${arg}}")
```

Comment:
@tejlmand: https://github.com/zephyrproject-rtos/zephyr/pull/57884#discussion_r1200522112

> it just seems more overhead to keep track of both included and excluded.

It's convenient though.

For example, you can use `INCLUDE` to define a group of bootloaders, but what if you also want to manage non-bootloaders?
```
ExternalZephyrProject_Group(
  non_bootloaders
  INCLUDE_FROM all
  EXCLUDE_FROM bootloaders
)
```

Reply:
Also seems like a good way to keep track of boards, e.g. have groups for each board, then you can select all images for each board and exclude the rest

Reply:
Fair enough.
Let's keep it.